src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,536 @@
+            idx = psy_data_nodes.index(self)
+            region_name += ":r{0}".format(idx)
+
+        if not options:
+            options = {}
+
+        pre_variable_list = options.get("pre-var-list", [])
+        post_variable_list = options.get("post-var-list", [])
+        pre_suffix = options.get("pre-var-postfix", "")
+        post_suffix = options.get("post-var-postfix", "")
+
+        # Note that adding a use statement makes sure it is only
+        # added once, so we don't need to test this here!
+        use = UseGen(parent, self.fortran_module, only=True,
+                     funcnames=PSyDataNode.symbols)
+        parent.add(use)
+        var_decl = TypeDeclGen(parent, datatype="PSyDataType",
+                               entity_decls=[self._var_name],
+                               save=True)
+        parent.add(var_decl)
+
+        self._add_call("PreStart", parent,
+                       ["\"{0}\"".format(module_name),
+                        "\"{0}\"".format(region_name),
+                        len(pre_variable_list),
+                        len(post_variable_list)])
+        self.set_region_identifier(module_name, region_name)
+        has_var = pre_variable_list or post_variable_list
+
+        # All variable names get a postfix of "-post" or "-pr".
+        # This enables the netcdf file to store variables that
+        # read read-write (i.e. pre and post), and also avoids
+        # potential name clashes (e.g. when using the PSyData
+        # interface to create a netcdf file), since a Fortran
+        # variable name cannot contain a "-".
+        if has_var:
+            for var_name in pre_variable_list:
+                self._add_call("PreDeclareVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, pre_suffix),
+                                "{0}".format(var_name)])
+            for var_name in post_variable_list:
+                self._add_call("PreDeclareVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, post_suffix),
+                                "{0}".format(var_name)])
+
+            self._add_call("PreEndDeclaration", parent)
+
+            for var_name in pre_variable_list:
+                self._add_call("ProvideVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, pre_suffix),
+                                "{0}".format(var_name)])
+
+            self._add_call("PreEnd", parent)
+
+        for child in self.psy_data_body:
+            child.gen_code(parent)
+
+        if has_var:
+            # Only add PostStart() if there is at least one variable.
+            self._add_call("PostStart", parent)
+            for var_name in post_variable_list:
+                self._add_call("ProvideVariable", parent,
+                               ["\"{0}{1}\"".format(var_name, post_suffix),
+                                "{0}".format(var_name)])
+
+        self._add_call("PostEnd", parent)
```

Comment:
So we always call PostEnd() but not always PostStart()?

Reply:
Correct - I could as easily have always created PostStart(), and PostEnd only when variable where there ... but it felt somehow better (for profiling API) if I had a call with 'End' at the end, not with Start